pkg/file/tar_index.go

```diff
@@ -0,0 +1,70 @@
+package file
+
+import (
+	"fmt"
+	"io"
+	"os"
+)
+
+type TarIndexVisitor func(TarIndexEntry) error
+
+// TarIndex is a tar reader capable of O(1) fetching of entry contents after the first read.
+type TarIndex struct {
+	indexByName map[string][]TarIndexEntry
+}
+
+// NewTarIndex creates a new TarIndex that is already indexed.
+func NewTarIndex(tarFilePath string, onIndex ...TarIndexVisitor) (*TarIndex, error) {
+	t := &TarIndex{
+		indexByName: make(map[string][]TarIndexEntry),
+	}
+	fh, err := os.Open(tarFilePath)
+	if err != nil {
+		return nil, err
+	}
+	defer fh.Close()
+	return t, t.indexEntries(fh, onIndex...)
```

Comment:
I'm not used to seeing an object's method be used in its own construction in OOP, but I like that the indexing functionality was broken out into its own function. Is there something that necessitated this approach?

Reply:
when this was a `IndexedTarReader` implementation it made a lot of sense, but with the smaller implementation as a `TarIndex` it doesn't make much sense --I'll merge both of these functions together.